File: DFS&BFS/1_DFS/dfs.py
def dfs(graph, v, visited):
    visited[v] = True
    print(v, end=' ')
    for i in graph[v]:
        if not visited[i]:
            dfs(graph, i, visited)


graph = [
    [],  # 0번 인덱스는 진입 노드가 1번부터 시작하기 때문에 비워놓는다.
    [2, 3, 8],
    [1, 7],
    [1, 4, 5],
    [3, 5],
    [3, 4],
    [7],
    [2, 6, 8],
    [1, 7]
]

# 노드 크기만큼만 만들어서 각 인덱스 -1할 수도 있지만 하나 더 크게 만들고 0번 인덱스 비우는 편이 덜 헷갈린다.
visited = [False] * 9

dfs(graph, 1, visited)

File: DFS&BFS/1_DFS/dfs.cpp
#include <bits/stdc++.h>

using namespace std;

bool visited[9];
vector<int> graph[9];

void dfs(int x)
{
    visited[x] = true;
    cout << x << ' ';
    for (int i = 0; i < graph[x].size(); i++)
    {
        int y = graph[x][i];
        if (!visited[y])
            dfs(y);
    }
}

int main(void)
{
    graph[0].push_back(0);
    graph[1].push_back(2);
    graph[1].push_back(3);
    graph[1].push_back(8);
    graph[2].push_back(1);
    graph[2].push_back(7);
    graph[3].push_back(1);
    graph[3].push_back(4);
    graph[3].push_back(5);
    graph[4].push_back(3);
    graph[4].push_back(5);
    graph[5].push_back(3);
    graph[5].push_back(4);
    graph[6].push_back(7);
    graph[7].push_back(2);
    graph[7].push_back(6);
    graph[7].push_back(8);
    graph[8].push_back(1);
    graph[8].push_back(7);

    visited[0] = false;
    visited[1] = false;
    visited[2] = false;
    visited[3] = false;
    visited[4] = false;
    visited[5] = false;
    visited[6] = false;
    visited[7] = false;
    visited[8] = false;

    dfs(1);
}